src/db.c

```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
   db_query_run("UPDATE speakers SET selected = 0;", 0, 0);
 }
 
+/* Queue */
 
-/* Inotify */
-int
-db_watch_clear(void)
-{
-  return db_query_run("DELETE FROM inotify;", 0, 0);
+static int
+queue_fetch_byitemid(struct db_queue_enum *queue_enum, uint32_t item_id, struct db_queue_item *queue_item, int keep_item);
+
+
+void
+free_queue_item(struct db_queue_item *queue_item, int content_only)
+{
+  if (queue_item->path)
+    free(queue_item->path);
+  if (queue_item->virtual_path)
+    free(queue_item->virtual_path);
+  if (queue_item->title)
+    free(queue_item->title);
+  if (queue_item->artist)
+    free(queue_item->artist);
+  if (queue_item->album_artist)
+    free(queue_item->album_artist);
+  if (queue_item->album)
+    free(queue_item->album);
+  if (queue_item->genre)
+    free(queue_item->genre);
+  if (queue_item->artist_sort)
+    free(queue_item->artist_sort);
+  if (queue_item->album_sort)
+    free(queue_item->album_sort);
+  if (queue_item->album_artist_sort)
+    free(queue_item->album_artist_sort);
+
+  if (!content_only && queue_item)
+    free(queue_item);
 }
 
+/*
+ * Returns the queue version from the admin table
+ *
+ * @return queue version
+ */
 int
-db_watch_add(struct watch_info *wi)
+db_queue_get_version()
 {
-#define Q_TMPL "INSERT INTO inotify (wd, cookie, path) VALUES (%d, 0, '%q');"
-  char *query;
+  char *version;
+  int32_t plversion;
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wi->wd, wi->path);
+  plversion = 0;
+  version = db_admin_get("plversion");
+  if (version)
+    {
+      ret = safe_atoi32(version, &plversion);
+      free(version);
+      if (ret < 0)
+	{
+	  DPRINTF(E_LOG, L_DB, "Could not get playlist version\n");
+	  return -1;
+	}
+    }
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
+  return plversion;
 }
 
-int
-db_watch_delete_bywd(uint32_t wd)
+/*
+ * Increments the version of the queue in the admin table and notifies listener of LISTENER_PLAYLIST
+ * about the change.
+ *
+ * This function must be called after successfully modifying the queue table in order to send
+ * notification messages to the clients (e. g. dacp or mpd clients).
+ */
+static void
+queue_inc_version_and_notify()
 {
-#define Q_TMPL "DELETE FROM inotify WHERE wd = %d;"
-  char *query;
+  int plversion;
+  char version[10];
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wd);
+  db_transaction_begin();
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
-}
+  plversion = db_queue_get_version();
```

Comment:
I would avoid custom sqlite extensions as long as they do not really add a benefit. If incrementing a value stored as an int is used some place else in forked-daapd, that may be the case. Otherwise i think the sql extension would make it harder to understand what happens.